src/memo/memo.ml

```diff
@@ -1,31 +1,97 @@
 open! Stdune
 open Fiber.O
 
-module type Input = Memo_intf.Input
-module type Output = Memo_intf.Output
-module type Decoder = Memo_intf.Decoder
-
 module Function_name = Interned.Make(struct
     let initial_size = 1024
     let resize_policy = Interned.Greedy
     let order = Interned.Fast
   end) ()
 
+module Function_type = struct
+  type ('a, 'b, 'f) t =
+    | Sync : ('a, 'b, ('a -> 'b)) t
+    | Async : ('a, 'b, ('a -> 'b Fiber.t)) t
+end
+
+module Function = struct
+  type ('a, 'b, 'f) t =
+    | Sync : ('a -> 'b) -> ('a, 'b, ('a -> 'b)) t
+    | Async : ('a -> 'b Fiber.t) -> ('a, 'b, ('a -> 'b Fiber.t)) t
+
+  let of_type
+        (type a) (type b) (type f) (t : (a, b, f) Function_type.t) (f : f) : (a, b, f) t =
+    match t with
+    | Function_type.Sync -> Sync f
+    | Function_type.Async -> Async f
+end
+
+module Witness : sig
+  type 'a t
+
+  val create : unit -> 'a t
+
+  val same : 'a t -> 'b t -> ('a, 'b) Type_eq.t option
+end = struct
+
+  type _ w = ..
+
+  module type T = sig
+    type a
+    type _ w += W : a w
+  end
+
+  type 'a t = (module T with type a = 'a)
+
+  let create (type a) () =
+    ((module struct
+      type nonrec a = a
+      type _ w += W : a w
+    end) : a t)
+
+  let same (type a) (type b) ((module M1) : a t) ((module M2) : b t) =
+    match M1.W with
+    | M2.W -> Some (Type_eq.T : (a, b) Type_eq.t)
+    | _ -> None
+
+end
+
+module Allow_cutoff = struct
+  type 'o t =
+    | No
+    | Yes of { equal : 'o -> 'o -> bool }
```

Comment:
Inline records aren't allowed in dune because we still support 4.02 unfortunately

Reply:
Whoops. Fixed now.